### DISCUSSION TITLE
Open workspace format modification

### DISCUSSION THREAD

**Discussion by ezhilkarthi** [eclipse-theia/theia#9873] [General] [answered]
Hi All,

I wish to remove the default format "All Files" and add a custom format as the default.
ex: (Instead of "All Files" tsx should be default)
![image](https://user-images.githubusercontent.com/31158252/128874866-08e4024c-ebbf-4a33-a60e-52396be65fd6.png)

**Comment by vince-fugnitto** [comment]
@ezhilkarthi I don't believe it is possible to completely remove `all files` filter yet, but you can add a default filter of `.tsx` if necessary.

If you aim to modify the existing `open workspace` dialog you'll need to create a [custom theia extension](https://theia-ide.org/docs/authoring_extensions/), and modify the `props` passed to the dialog (you're interested in `filters`):

https://github.com/eclipse-theia/theia/blob/19ace46f5dadc3345cc14a7edf9f85a9a025a9b0/packages/workspace/src/browser/workspace-frontend-contribution.ts#L351-L376

https://github.com/eclipse-theia/theia/blob/19ace46f5dadc3345cc14a7edf9f85a9a025a9b0/packages/workspace/src/browser/workspace-frontend-contribution.ts#L531-L584

You would `rebind` (using inversify's dependency injection) the existing functionality to your custom implementation.

**Reply by ezhilkarthi** [reply]
Hi @vince-fugnitto,
Thanks, I can understand. As you can see in the screenshot below all files is selected as default, can you suggest me a way to select the "tsx" file as the default
![image](https://user-images.githubusercontent.com/31158252/128878421-85b92471-3ae7-40be-b018-47a7f1e5e608.png)

**Reply by msujew** [reply]
Which Theia version are you using? This should be fixed as of 1.15.0 with #9659.

**Reply by ezhilkarthi** [reply]
Hi @msujew,
I am using "@theia/core": "~1.13.0". I will look into this.
Thanks for the update.

**Reply by ezhilkarthi** [reply]
Hi @msujew @vince-fugnitto, I updated to 1.17.2. While initial loading it was selected TSX but the filter is not happening. Instead of that, it showing all the files. Once we select another file only filter was applied

**Reply by vince-fugnitto** [reply]
@ezhilkarthi do you have an example we can reference to see the behavior, is your extension public?

**Reply by ezhilkarthi** [reply]
@vince-fugnitto, I believe the filter isn't applied on loading. The filter only works if the selection is changed. This can be because the handleFilterChanged isn't called on initialization within the constructor as shown in the screenshot. 
![image](https://user-images.githubusercontent.com/31158252/133317073-385011c3-c66d-46a2-87f9-e6ee5aacb0e1.png)

**Reply by vince-fugnitto** [reply]
@ezhilkarthi how are you modifying the filter in your custom extension, or creating the `FileDialog`?

**Reply by ezhilkarthi** [reply]
Hi @vince-fugnitto, I'm currently following this code. Do let me know if I need to change something.

```
` const openProjectDialogProps: OpenFileDialogProps = {
      title: "Open Project File",
      filters: { "tsx": [tsx]},
      canSelectFolders: false,
      canSelectMany: false,
      openLabel: "Open",
    };
const defaultURI = await fileService.resolve(new URI(getProjectPath()))
this.fileDialogService.showOpenDialog(openProjectDialogProps, defaultURI);
```
`
Thanks

**Reply by vince-fugnitto** [reply]
@ezhilkarthi it worked well for me when I tried (both the browser and electron):

![image](https://user-images.githubusercontent.com/40359487/133781234-4cc1733e-8e1f-4168-acfd-ecd133f5ba8d.png)


I kept the example simple:

```ts
protected async openDialog(): Promise<void> {
    await this.fileDialogService.showOpenDialog({
        title: 'Getting Started',
        filters: {
            'TypeScript': ['ts', 'tsx']
        }
    });
}
```

The `TypeScript` filter was set first, and `All Files` last.

**Reply by ezhilkarthi** [reply]
@vince-fugnitto, TypeScript was at first but TS or TSX files should only be visible in dialog

**Reply by ezhilkarthi** [reply]
@vince-fugnitto, I choose typescript, but It showing all the files like HTML, JS...

![image](https://user-images.githubusercontent.com/31158252/133794366-0747649d-9b3c-485c-84f5-72b77aa86d18.png)


`
 ```
 const openProjectDialogProps: OpenFileDialogProps = {
      title: "Open Project File",
      filters: { "TypeScript": ['ts', 'tsx'] },
      canSelectFolders: false,
      canSelectMany: false,
      openLabel: "Open",
    };
```

`

**Reply by vince-fugnitto** [reply]
@ezhilkarth unfortunately, I'm not sure I follow anymore.

If you wish to remove 'all files' you'll need to do so in a custom extension like I've mentioned in https://github.com/eclipse-theia/theia/discussions/9873#discussioncomment-1154366.

In https://github.com/eclipse-theia/theia/discussions/9873#discussioncomment-1154431 you mentioned having TypeScript as the default which is the case now as you've highlighted. In my example when the TypeScript filter is selected only TypeScript files are visible and can be opened.

**Reply by vince-fugnitto** [reply]
@ezhilkarthi I believe I might see what you see when doing so in the browser, the dialog does not filter out the resources which do not match the filter.

**Reply by ezhilkarthi** [reply]
@vince-fugnitto, Yes, the issue is with the filtering. On the dialog box being opened, it doesn't filter the files. However, on selecting all files and coming back to the default filter i.e tsx one, it's filtering.

**Reply by ezhilkarthi** [reply]
@vince-fugnitto, Is there any workaround in the current implementation, I am using theia 1.17.2.

**Reply by vince-fugnitto** [reply]
@ezhilkarthi it will likely need to be fixed, I'll open a follow-up issue. In electron the use-case works well.

**Comment by vince-fugnitto** [comment, marked as answer]
@ezhilkarthi the issue you've experienced before should be fixed with https://github.com/eclipse-theia/theia/pull/10133.
Again, thank you for reporting it :+1:
